File: 351002/Stepanenko/src/main/java/by/bsuir/distcomp/service/TweetService.java
package by.bsuir.distcomp.service;

import by.bsuir.distcomp.dto.request.TweetRequestTo;
import by.bsuir.distcomp.dto.response.TweetResponseTo;
import by.bsuir.distcomp.entity.Tweet;
import by.bsuir.distcomp.exception.ResourceNotFoundException;
import by.bsuir.distcomp.mapper.TweetMapper;
import by.bsuir.distcomp.repository.impl.InMemoryAuthorRepository;
import by.bsuir.distcomp.repository.impl.InMemoryMarkerRepository;
import by.bsuir.distcomp.repository.impl.InMemoryTweetRepository;
import org.springframework.stereotype.Service;

import java.time.LocalDateTime;
import java.util.HashSet;
import java.util.List;
import java.util.Set;
import java.util.stream.Collectors;

@Service
public class TweetService {
    private final InMemoryTweetRepository tweetRepository;
    private final InMemoryAuthorRepository authorRepository;
    private final InMemoryMarkerRepository markerRepository;
    private final TweetMapper tweetMapper;

    public TweetService(InMemoryTweetRepository tweetRepository,
                        InMemoryAuthorRepository authorRepository,
                        InMemoryMarkerRepository markerRepository,
                        TweetMapper tweetMapper) {
        this.tweetRepository = tweetRepository;
        this.authorRepository = authorRepository;
        this.markerRepository = markerRepository;
        this.tweetMapper = tweetMapper;
    }

    public TweetResponseTo create(TweetRequestTo dto) {
        if (!authorRepository.existsById(dto.getAuthorId())) {
            throw new ResourceNotFoundException("Author with id " + dto.getAuthorId() + " not found", 40404);
        }
        
        if (dto.getMarkerIds() != null && !dto.getMarkerIds().isEmpty()) {
            for (Long markerId : dto.getMarkerIds()) {
                if (!markerRepository.existsById(markerId)) {
                    throw new ResourceNotFoundException("Marker with id " + markerId + " not found", 40405);
                }
            }
        }
        
        Tweet entity = tweetMapper.toEntity(dto);
        entity.setCreated(LocalDateTime.now());
        entity.setModified(LocalDateTime.now());
        if (dto.getMarkerIds() != null) {
            entity.setMarkerIds(dto.getMarkerIds());
        }
        Tweet saved = tweetRepository.save(entity);
        return tweetMapper.toResponseDto(saved);
    }

    public TweetResponseTo getById(Long id) {
        Tweet entity = tweetRepository.findById(id)
                .orElseThrow(() -> new ResourceNotFoundException("Tweet with id " + id + " not found", 40406));
        return tweetMapper.toResponseDto(entity);
    }

    public List<TweetResponseTo> getAll() {
        return tweetRepository.findAll().stream()
                .map(tweetMapper::toResponseDto)
                .collect(Collectors.toList());
    }

    public TweetResponseTo update(TweetRequestTo dto) {
        Tweet existingTweet = tweetRepository.findById(dto.getId())
                .orElseThrow(() -> new ResourceNotFoundException("Tweet with id " + dto.getId() + " not found", 40407));
        
        if (!authorRepository.existsById(dto.getAuthorId())) {
            throw new ResourceNotFoundException("Author with id " + dto.getAuthorId() + " not found", 40408);
        }
        
        if (dto.getMarkerIds() != null && !dto.getMarkerIds().isEmpty()) {
            for (Long markerId : dto.getMarkerIds()) {
                if (!markerRepository.existsById(markerId)) {
                    throw new ResourceNotFoundException("Marker with id " + markerId + " not found", 40409);
                }
            }
        }
        
        Tweet entity = tweetMapper.toEntity(dto);
        entity.setCreated(existingTweet.getCreated());
        entity.setModified(LocalDateTime.now());
        if (dto.getMarkerIds() != null) {
            entity.setMarkerIds(dto.getMarkerIds());
        } else {
            entity.setMarkerIds(existingTweet.getMarkerIds());
        }
        Tweet updated = tweetRepository.update(entity);
        return tweetMapper.toResponseDto(updated);
    }

    public void deleteById(Long id) {
        if (!tweetRepository.existsById(id)) {
            throw new ResourceNotFoundException("Tweet with id " + id + " not found", 40410);
        }
        tweetRepository.deleteById(id);
    }

    public TweetResponseTo getAuthorByTweetId(Long tweetId) {
        Tweet tweet = tweetRepository.findById(tweetId)
                .orElseThrow(() -> new ResourceNotFoundException("Tweet with id " + tweetId + " not found", 40411));
        return getById(tweetId);
    }

    public List<TweetResponseTo> getTweetsByMarkerNames(Set<String> markerNames,
                                                         Set<Long> markerIds,
                                                         String authorLogin,
                                                         String title,
                                                         String content) {
        List<Tweet> tweets = tweetRepository.findAll();
        
        Set<Long> finalMarkerIds = markerIds;
        if (markerNames != null && !markerNames.isEmpty()) {
            Set<Long> markerIdsByName = markerRepository.findAll().stream()
                    .filter(m -> markerNames.contains(m.getName()))
                    .map(m -> m.getId())
                    .collect(Collectors.toSet());
            if (finalMarkerIds == null) {
                finalMarkerIds = markerIdsByName;
            } else {
                finalMarkerIds = new HashSet<>(finalMarkerIds);
                finalMarkerIds.addAll(markerIdsByName);
            }
        }
        
        final Set<Long> markerIdsForFilter = finalMarkerIds;
        if (markerIdsForFilter != null && !markerIdsForFilter.isEmpty()) {
            tweets = tweets.stream()
                    .filter(t -> t.getMarkerIds() != null && 
                            t.getMarkerIds().stream().anyMatch(markerIdsForFilter::contains))
                    .collect(Collectors.toList());
        }
        
        if (authorLogin != null && !authorLogin.isEmpty()) {
            Set<Long> authorIds = authorRepository.findAll().stream()
                    .filter(a -> authorLogin.equals(a.getLogin()))
                    .map(a -> a.getId())
                    .collect(Collectors.toSet());
            tweets = tweets.stream()
                    .filter(t -> authorIds.contains(t.getAuthorId()))
                    .collect(Collectors.toList());
        }
        
        if (title != null && !title.isEmpty()) {
            tweets = tweets.stream()
                    .filter(t -> title.equals(t.getTitle()))
                    .collect(Collectors.toList());
        }
        
        if (content != null && !content.isEmpty()) {
            tweets = tweets.stream()
                    .filter(t -> t.getContent() != null && t.getContent().contains(content))
                    .collect(Collectors.toList());
        }
        
        return tweets.stream()
                .map(tweetMapper::toResponseDto)
                .collect(Collectors.toList());
    }
}

File: 351003/Guzaev/app/services/tweet_service.py
from app.repositories.base_repository import InMemoryRepository
from app.models.tweet import Tweet
from app.dtos.tweet_dto import TweetRequestTo, TweetResponseTo
from app.services.writer_service import writer_repository
from app.services.marker_service import marker_repository
from app.errors import AppError
from typing import List
from datetime import datetime

tweet_repository = InMemoryRepository[Tweet]()


class TweetService:
    def create(self, dto: TweetRequestTo) -> TweetResponseTo:
        # 1. Проверка существования Writer
        writer = writer_repository.get_by_id(dto.writer_id)
        if not writer:
            raise AppError(status_code=403, message="Writer not found", error_code=40303)

        # 2. Проверка существования маркеров
        if dto.marker_ids:
            for marker_id in dto.marker_ids:
                marker = marker_repository.get_by_id(marker_id)
                if not marker:
                    raise AppError(status_code=403, message=f"Marker {marker_id} not found", error_code=40304)

        # 3. Создание сущности
        entity = Tweet(
            writer_id=dto.writer_id,
            title=dto.title,
            content=dto.content,
            marker_ids=dto.marker_ids or [],
            created=datetime.now(),
            modified=datetime.now()
        )
        saved_entity = tweet_repository.create(entity)

        # 4. Явный маппинг в DTO (исправление ошибки 500)
        return TweetResponseTo(
            id=saved_entity.id,
            writerId=saved_entity.writer_id,
            title=saved_entity.title,
            content=saved_entity.content,
            created=saved_entity.created,
            modified=saved_entity.modified,
            markerIds=saved_entity.marker_ids
        )

    def get_all(self) -> List[TweetResponseTo]:
        entities = tweet_repository.get_all()
        return [
            TweetResponseTo(
                id=e.id,
                writerId=e.writer_id,
                title=e.title,
                content=e.content,
                created=e.created,
                modified=e.modified,
                markerIds=e.marker_ids
            ) for e in entities
        ]

    def get_by_id(self, id: int) -> TweetResponseTo:
        entity = tweet_repository.get_by_id(id)
        if not entity:
            raise AppError(status_code=404, message="Tweet not found", error_code=40404)

        return TweetResponseTo(
            id=entity.id,
            writerId=entity.writer_id,
            title=entity.title,
            content=entity.content,
            created=entity.created,
            modified=entity.modified,
            markerIds=entity.marker_ids
        )

    def update(self, id: int, dto: TweetRequestTo) -> TweetResponseTo:
        entity = tweet_repository.get_by_id(id)
        if not entity:
            raise AppError(status_code=404, message="Tweet not found", error_code=40405)

        # Проверка Writer
        writer = writer_repository.get_by_id(dto.writer_id)
        if not writer:
            raise AppError(status_code=403, message="Writer not found", error_code=40305)

        # Проверка Markers
        if dto.marker_ids:
            for marker_id in dto.marker_ids:
                marker = marker_repository.get_by_id(marker_id)
                if not marker:
                    raise AppError(status_code=403, message=f"Marker {marker_id} not found", error_code=40306)

        # Обновление
        updated_entity = Tweet(
            id=id,
            writer_id=dto.writer_id,
            title=dto.title,
            content=dto.content,
            marker_ids=dto.marker_ids or [],
            created=entity.created,  # дата создания не меняется
            modified=datetime.now()  # дата обновления меняется
        )
        tweet_repository.update(id, updated_entity)

        return TweetResponseTo(
            id=updated_entity.id,
            writerId=updated_entity.writer_id,
            title=updated_entity.title,
            content=updated_entity.content,
            created=updated_entity.created,
            modified=updated_entity.modified,
            markerIds=updated_entity.marker_ids
        )

    def delete(self, id: int) -> None:
        entity = tweet_repository.get_by_id(id)
        if not entity:
            raise AppError(status_code=404, message="Tweet not found", error_code=40406)
        tweet_repository.delete(id)
